File: ccc23s3.cpp
#include <iostream>

char P[2000][2000];
int n, m;

using namespace std;

void print() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) cout << P[i][j];
        cout << "\n";
    }
}

int main() {
    int r, c; cin >> n >> m >> r >> c;
    if ((n == r && m % 2 == 0 && c % 2 == 1) || (m == c && n % 2 == 0 && r % 2 == 1)) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }
    for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) P[i][j] = 'a';
    if (r == n && c != m) {
        for (int i = 0; i < m; i++) P[0][i] = 'b';
        if (c % 2 == 1) P[0][m / 2] = 'a';
        for (int i = 0; i < c / 2; i++) {
            P[0][i] = 'a';
            P[0][m - i - 1] = 'a';
        }
        print();
        return 0;
    } else if (c == m && r != n) {
        for (int i = 0; i < n; i++) P[i][0] = 'b';
        if (r % 2 == 1) P[n / 2][0] = 'a';
        for (int i = 0; i < r / 2; i++) {
            P[i][0] = 'a';
            P[n - i - 1][0] = 'a';
        }
        print();
        return 0;
    }
    if (n != r) for (int i = 0; i < n - r; i++) P[i][0] = 'b';
    if (m != c) for (int i = 0; i < m - c; i++) P[0][i] = 'b';
    if (n - r != 0 && m - c != 0) P[0][0] = 'c';
    print();
    return 0;
}

File: ccc23s3.py
import sys

def print_poster(P):
    for i in P:
        print(''.join(i))


n, m, r, c = map(int, input().split())

if (n == r and m % 2 == 0 and c % 2 == 1) or (m == c and n % 2 == 0 and r % 2 == 1):
    print("IMPOSSIBLE")
    sys.exit()

P = [['a' for _ in range(m)] for _ in range(n)]
if r == n and c != m:
    for i in range(m):
        P[0][i] = 'b'
    if c % 2 == 1:
        P[0][m // 2] = 'a'
    for i in range(c // 2):
        P[0][i] = 'a'
        P[0][m - i - 1] = 'a'
    print_poster(P)
    sys.exit()
elif c == m and r != n:
    for i in range(n):
        P[i][0] = 'b'
    if r % 2 == 1:
        P[n // 2][0] = 'a'
    for i in range(r // 2):
        P[i][0] = 'a'
        P[n - i - 1][0] = 'a'
    print_poster(P)
    sys.exit()

if n != r:
    for i in range(n - r):
        P[i][0] = 'b'

if m != c:
    for i in range(m - c):
        P[0][i] = 'b'

if n - r != 0 and m - c != 0:
    P[0][0] = 'c'

print_poster(P)